proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+  bogusCombinators: Deprecation<'bogus-combinators', 'active'>;
+
+  /** Deprecation for ambiguous `+` and `-` operators. */
+  strictUnary: Deprecation<'strict-unary', 'active'>;
+
+  /** Deprecation for passing invalid units to certain built-in functions. */
+  functionUnits: Deprecation<'function-units', 'active'>;
+
+  /** Deprecation for `@import` rules. */
+  importRules: Deprecation<'import', 'future'>;
+
+  /** Used for deprecations coming from user-authored code. */
+  userAuthored: Deprecation<'user-authored', 'user'>;
+
+  /**
+   * Returns the deprecation with the given ID, or null if none exists.
+   *
+   * This must support all specified deprecations, but implementations may also
+   * have their own deprecations beyond those specified by the language.
+   */
+  find(id: string): Deprecation<typeof id, DeprecationStatus> | null;
+}
+
+/** A deprecation, or the ID of one. */
+export type DeprecationOrId = Deprecation | string;
+
+/** A deprecation's status. */
+export type DeprecationStatus = 'active' | 'user' | 'future' | 'obsolete';
+
+/** A deprecated feature in the language. */
+export interface Deprecation<
+  id extends string = string,
+  status extends DeprecationStatus = DeprecationStatus
+> {
+  /**
+   * An ID for this deprecation.
+   *
+   * For language-wide deprecations specified in the API, this will be a short,
+   * alphanumeric, kebab-case identifier. Implementation-specific deprecations
+   * should be prefixed with an implementation identifier followed by a slash to
+   * avoid collision with future language-wide deprecations.
+   */
+  id: id;
+
+  /**
+   * The status of this deprecation.
+   *
+   * - 'active' means this deprecation is currently enabled. `deprecatedIn` is
+   *   non-null and `obsoleteIn` is null.
+   * - 'user' means this deprecation is from user-authored code. Both
+   *   `deprecatedIn` and `obsoleteIn` are null.
+   * - 'future' means this deprecation is not yet enabled. Both `deprecatedIn`
+   *   and `obsoleteIn` are null.
+   * - 'obsolete' means this deprecation is now obsolete, as the feature it was
+   *   for has been fully removed. Both `deprecatedIn` and `obsoleteIn` are
+   *   non-null.
+   */
+  status: status;
+
+  /** A brief user-readable description of this deprecation. */
+  description?: string;
+
+  /**
+   * The compiler version this feature was first deprecated in.
+   *
+   * This is implementation-dependent, so versions are not guaranteed to be
+   * consistent between different compilers. For future deprecations, or those
+   * originating from user-authored code, this is null.
+   */
+  deprecatedIn: status extends 'future' | 'user' ? null : Version;
+
+  /**
+   * The compiler version this feature was fully removed in, making the
+   * deprecation obsolete.
+   *
+   * This is null for active and future deprecations.
+   */
+  obsoleteIn: status extends 'obsolete' ? Version : null;
+}
+
+/** A semantic version of the compiler. */
+export interface Version {
```

Comment:
I'm concerned that this custom version struct will be annoying for users to construct by hand relative to just writing out the version number. It's probably not worth depending on the JS `semver` library since it's not ES-compatible and it has its own transitive dependencies, but maybe it would be worth making this a class with a constructor that can parse a string.